### PR TITLE
Flink 5386 pr inner

### PR DESCRIPTION
**Pull request by sunjincheng121**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [ ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala [hunk 1]
```diff
@@ -66,6 +66,15 @@ class Table(
   private val tableSchema: TableSchema = new TableSchema(
     logicalPlan.output.map(_.name).toArray,
     logicalPlan.output.map(_.resultType).toArray)
+  
+  // map to save the all windows information registered in the table
+  private[flink] var windowPool: Map[Expression, GroupWindow] = Map()
+
+  // get the windowPool from the existing table
```
Comment (lincoln-lil) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala]: better using /** xxxx */ for single line comment of a function

see http://docs.scala-lang.org/style/scaladoc.html

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/windows.scala [hunk 2]
```diff
@@ -34,8 +34,9 @@ import org.apache.flink.table.plan.logical._
   * For finite batch tables, group-windows provide shortcuts for time-based groupBy.
   *
   */
-trait GroupWindow {
+abstract class GroupWindow {
 
+  private[flink] var alias: Option[Expression] = None
```
Comment (shaoxuan-wang) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/windows.scala]: 加上注释 "The expression of alias for this GroupWindow"

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala (hunk 1 above)
Comment (lincoln-lil) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala]: return type of this builder function is a Table, and parameter type is also a Table, so
this function does "withWindowPoolFromGivenTable", and comment as "set windowPool of the Table from an existing Table Instance" maybe more precisely
the short function name "withWindowPool" is also good for me

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala [hunk 3]
```diff
@@ -855,78 +919,46 @@ class GroupedTable(
...
-      fields, table.tableEnv, aggNames, propNames)
-
-    val projectFields = (table.tableEnv, window) match {
-      // event time can be arbitrary field in batch environment
-      case (_: BatchTableEnvironment, w: EventTimeWindow) =>
-        extractFieldReferences(fields ++ groupKey ++ Seq(w.timeField))
-      case (_, _) =>
-        extractFieldReferences(fields ++ groupKey)
+  def groupBy(fields: Expression*): GroupedTable = {
+    if (fields.filter(table.windowPool.contains(_)).length > 1) {
```
Comment (shaoxuan-wang) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala]: windowedTable.groupBy 之后不应该是windowedGroupbyTable?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala [hunk 4]
```diff
@@ -66,6 +66,15 @@ class Table(
   private val tableSchema: TableSchema = new TableSchema(
     logicalPlan.output.map(_.name).toArray,
     logicalPlan.output.map(_.resultType).toArray)
+  
+  // map to save the all windows information registered in the table
```
Comment (lincoln-lil) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala]: "a Map stores all concrete Windows registered in the Table instance"
seems better?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala [hunk 5]
```diff
@@ -855,78 +919,46 @@ class GroupedTable(
...
-        projectsOnAgg,
-        WindowAggregate(
-          groupKey,
-          window.toLogicalWindow,
-          propNames.map(a => Alias(a._1, a._2)).toSeq,
-          aggNames.map(a => Alias(a._1, a._2)).toSeq,
-          Project(projectFields, table.logicalPlan).validate(table.tableEnv)
-        ).validate(table.tableEnv)
-      ).validate(table.tableEnv))
+    new GroupedTable(table, fields)
```
Comment (shaoxuan-wang) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala]: 你可以拓展一下GroupedTable接口，把windowAlias放进去就不需要withWindowList了
```
class GroupedTable(
  private[flink] val table: Table,
  private[flink] val windowAlias: Expression,
  private[flink] val groupKey: Seq[Expression])
```
new GroupedTable(table, window.alias, fields)

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala [hunk 6]
```diff
@@ -799,10 +815,16 @@ class Table(
     * will be processed by a single operator.
     *
     * @param groupWindow group-window that specifies how elements are grouped.
-    * @return A windowed table.
     */
```
Comment (lincoln-lil) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala]: return type comment should not be removed

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala [hunk 7]
```diff
@@ -799,10 +815,16 @@ class Table(
...
-  def window(groupWindow: GroupWindow): GroupWindowedTable = {
-    new GroupWindowedTable(this, Seq(), groupWindow)
+  def window(groupWindow: GroupWindow): WindowedTable = {
+    if (None == groupWindow.alias) {
+      throw new ValidationException("An alias must be specified for the window.")
+    }
+    if (windowPool.contains(groupWindow.alias.get)) {
+      throw new ValidationException("The window alias can not be duplicated.")
```
Comment (lincoln-lil) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala]: indicating the duplicate name is better

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala [hunk 8]
```diff
@@ -855,78 +919,46 @@ class GroupedTable(
...
-class GroupWindowedTable(
+class WindowedTable(
     private[flink] val table: Table,
-    private[flink] val groupKey: Seq[Expression],
     private[flink] val window: GroupWindow) {
 
   /**
-    * Performs a selection operation on a windowed table. Similar to an SQL SELECT statement.
-    * The field expressions can contain complex expressions and aggregations.
+    * Groups the elements on some grouping keys. Use this before a selection with aggregations
```
Comment (lincoln-lil) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala]: -> "Groups the elements by some grouping keys. Using this ..."

**Pull request closed without merging**